py/client/pydeephaven/updateby.py

```diff
@@ -0,0 +1,239 @@
+
+    UNLIMITED = 0, table_pb2.MathContext.RoundingMode.HALF_UP
+    """unlimited precision arithmetic, rounding is half-up"""
+
+    DECIMAL32 = 7, table_pb2.MathContext.RoundingMode.HALF_EVEN
+    """a precision setting matching the IEEE 754R Decimal32 format, 7 digits, rounding is half-even"""
+
+    DECIMAL64 = 16, table_pb2.MathContext.RoundingMode.HALF_EVEN
+    """a precision setting matching the IEEE 754R Decimal64 format, 16 digits, rounding is half-even"""
+
+    DECIMAL128 = 34, table_pb2.MathContext.RoundingMode.HALF_EVEN
+    """a precision setting matching the IEEE 754R Decimal128 format, 34 digits, rounding is half-even"""
+
+
+class BadDataBehavior(Enum):
+    """An Enum defining ways to handle invalid data during update-by operations."""
+
+    RESET = table_pb2.BadDataBehavior.RESET
+    """Reset the state for the bucket to None when invalid data is encountered"""
+
+    SKIP = table_pb2.BadDataBehavior.SKIP
+    """Skip and do not process the invalid data without changing state"""
+
+    THROW = table_pb2.BadDataBehavior.THROW
+    """Throw an exception and abort processing when bad data is encountered"""
+
+    POISON = table_pb2.BadDataBehavior.POISON
+    """Allow the bad data to poison the result. This is only valid for use with NaN"""
+
+
+class OperationControl(UpdateByBase):
+    """A OperationControl represents control parameters for performing operations with the table
+        UpdateByOperation."""
+
+    def __init__(self, on_null: BadDataBehavior = BadDataBehavior.SKIP,
+                 on_nan: BadDataBehavior = BadDataBehavior.SKIP,
+                 big_value_context: MathContext = MathContext.DECIMAL128):
+        """Initializes an OperationControl for use with certain UpdateByOperation, such as EMAs.
+
+        Args:
+            on_null (BadDataBehavior): the behavior for when null values are encountered, default is SKIP
+            on_nan (BadDataBehavior): the behavior for when NaN values are encountered, default is SKIP
+            big_value_context (MathContext): the context to use when processing arbitrary precision numeric values,
+                default is DECIMAL128.
+        """
+        self.on_null = on_null
+        self.on_nan = on_nan
+        self.big_value_context = big_value_context
+
+    def make_grpc_request(self):
+        return GrpcUpdateByEmaOptions(on_null_value=self.on_null.value, on_nan_value=self.on_nan.value,
+                                      big_value_context=GrpcMathContext(precision=self.big_value_context.value[0],
+                                                                        rounding_mode=self.big_value_context.value[1]))
+
+
+class UpdateByOperation(UpdateByBase):
+    """A UpdateByOperation represents an operator for the Table update-by operation."""
+
+    def __init__(self, ub_column):
+        self.ub_column = ub_column
+
+    def make_grpc_request(self):
+        return GrpcUpdateByOperation(column=self.ub_column)
```

Comment:
type hint?
This also comes back to the question of if we want to expose the Grpc stuff at the top of the file.